client/network/src/protocol.rs

```diff
@@ -572,10 +572,11 @@ impl<B: BlockT, H: ExHashT> Protocol<B, H> {
 		self.context_data.peers.iter().map(|(id, peer)| (id, &peer.info))
 	}
 
-	pub fn on_custom_message(
+	fn on_custom_message(
 		&mut self,
 		who: PeerId,
 		data: BytesMut,
+		cx: &mut std::task::Context,
```

Comment:
Passing a context to an otherwise synchronous function gives the impression that this function is similar to a `poll` function that is repeatedly called. But `on_custom_message` is only called ... well on custom messages. I would much rather keep polling logic like `poll_block_announce_validation` within functions that adhere to the above convention.

Reply:
But that doesn't work, I would need to call `poll_block_announce_validation' as well in `on_custom_message`. So, either I revert my latest commit and do it as it was done before, or we keep this.

Reply:
We need to poll after `on_custom_message` or we don't register the future in the task.

Reply:
What I am suggesting in https://github.com/paritytech/substrate/pull/6958#issuecomment-700942807 is to ignore the fact that the future is not registered with the task for block announcements from the legacy substream for the following two reasons:

1. `Protocol:poll` is polled regularly and includes a `poll_block_announce_validation(cx)` at the very beginning.

2. This concerns block announcements send via the legacy substream only which (a) is not used much and (b) is going away soon.

As far as I can see this is a reasonable trade-off. What do you think @bkchr?

Reply:
Done